test/clientconn_state_transition_test.go

```diff
@@ -547,3 +550,82 @@ func awaitNoStateChange(ctx context.Context, t *testing.T, cc *grpc.ClientConn,
 		t.Fatalf("State changed from %q to %q when no state change was expected", currState, cc.GetState())
 	}
 }
+
+type testConnectivityStateSubscriber struct {
+	onMsgCh chan connectivity.State
+}
+
+func (ts *testConnectivityStateSubscriber) OnMessage(msg interface{}) {
+	ts.onMsgCh <- msg.(connectivity.State)
```

Comment:
A non-blocking write here to ensure that a test bug doesn't lead to the whole test package blocking and expiring the `7m` timeout?

Reply:
I want it to be blocking, but it needs to also block on the context.  Sadly these things are going to be really, really hard to avoid entirely.  Think about common things like `for x := range ch`, which I worked a bit to avoid here.  :(

Also this uncovered an issue that sometimes `cc.Close()` will block on the `OnMessage` callback.  This has to do with the locks taken in the pubsub package.  If `Stop()` is called after `Publish` schedules the message and its lock is released but before it is executed (and its lock is re-taken), then `Stop` will just set a bool and the callback will still be executed.

Maybe this is a bug?  I'll have to think about it.  But for now, calling `cc.Close()` in a goroutine works around it.